File: 0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }
    
    left := depth(root.Left)
    right := depth(root.Right)

    //fmt.Printf("%d->%d ", root.Val, abs(left - right))
    return abs(left - right) <= 1 && isBalanced(root.Left) && isBalanced(root.Right)
}

func depth(root *TreeNode) int {
    if root == nil {
        return 0
    }

    left := depth(root.Left)
    right := depth(root.Right)

    //fmt.Printf("%d->%d ", root.Val, max(left, right) + 1)
    return max(left, right) + 1
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return nodeHeight(root) != -1;
    }

private:
    int nodeHeight(TreeNode* root) {
        if (!root) {
            return 0;
        }
        
        int leftHeight = nodeHeight(root->left);
        if (leftHeight == -1) {
            return -1;
        }

        int rightHeight = nodeHeight(root->right);
        if (rightHeight == -1) {
            return -1;
        }

        //cout << root->val << "->" << abs(leftHeight - rightHeight) << " ";

        if (abs(leftHeight - rightHeight) > 1) {
            return -1;
        }

        return max(leftHeight, rightHeight) + 1;
    }
};